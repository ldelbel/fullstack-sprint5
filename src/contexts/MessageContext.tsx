import React from "react";
import { IMessageContext } from "../types/interfaces";

const MessageContext = React.createContext<IMessageContext>(
  {} as IMessageContext
);

export default MessageContext;
