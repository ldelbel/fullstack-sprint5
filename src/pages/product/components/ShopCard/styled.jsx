import styled from "styled-components";

export const StyledShopCard = styled.div`
  width: 470px;
  height: 255px;
  box-shadow: 0px 5px 5px -3px rgb(0 0 0 / 20%),
    0px 8px 10px 1px rgb(0 0 0 / 14%), 0px 3px 14px 2px rgb(0 0 0 / 12%);
  padding: 2rem 1.5rem;

  > p {
    text-align: center;
    color: #666;
    font-weight: 400;
    margin-bottom: 1.8rem;

    span {
      font-weight: bold;
      color: #4581a8;
      text-decoration: underline;
      cursor: pointer;
    }
  }

  .prices {
    display: flex;
    justify-content: space-between;
    margin-bottom: 1.7rem;

    strong {
      font-size: 1.6rem;
      color: #d0403a;
      display: flex;
      align-items: center;
    }

    > div {
      display: flex;
      flex-direction: column;
      color: #666;
      font-weight: 400;
      font-size: 0.7rem;

      span + span {
        margin-top: 0.4rem;
      }
    }
  }

  .addBtn {
    background: #008844;
    border: none;
    border-radius: 4px;
    width: 100%;
    color: white;
    font-weight: bold;
    font-size: 0.9rem;
    padding: 0.8rem 0;

    &:hover {
      cursor: pointer;
      box-shadow: 0 2px 4px #3b2e2e60;
    }
  }

  .deliver {
    display: flex;
    justify-content: space-between;
    margin-top: 1.5rem;

    &__info {
      display: flex;
      align-items: center;
      height: 2rem;

      span {
        font-size: 0.7rem;
        font-weight: 400;
        color: #333;
        margin-left: 0.5rem;

        &:hover {
          cursor: default;
        }
      }

      input {
        width: 31%;
        font-size: 1rem;
        padding: 0 0.5rem;
        height: 100%;
        border: 1px solid #bbbbbb;
        border-radius: 2px;
        margin-left: 0.5rem;

        &:hover {
          border: 1px solid #585858;
        }
      }

      button {
        height: 100%;
        border: 1px solid #bbbbbb;
        border-radius: 2px;
        background: #fff;
        margin-left: 0.5rem;
        color: #585858;

        &:hover {
          border: 1px solid #585858;
          cursor: pointer;
        }
      }
    }

    p {
      font-size: 0.6rem;
      white-space: nowrap;
      display: flex;
      align-items: center;
      color: #777;
      font-weight: 400;
      text-decoration: underline;

      &:hover {
        cursor: pointer;
        color: #555;
      }
    }
  }
`;
