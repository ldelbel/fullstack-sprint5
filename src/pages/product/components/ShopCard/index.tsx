import React, { ReactElement, useContext } from "react";
import { FiTruck } from "react-icons/fi";
import { useHistory } from "react-router-dom";
import MessageContext from "../../../../contexts/MessageContext";
import { StyledShopCard } from "./styled";

interface ShopCardProps {
  price: string;
}

export function ShopCard({ price }: ShopCardProps): ReactElement<React.FC> {
  const { setMessage } = useContext(MessageContext);
  const history = useHistory();

  const installment = () => {
    const valueDot = Number(price.replace(",", ".")) / 3;
    return valueDot.toFixed(2).replace(".", ",");
  };

  function handleAddToCart() {
    setMessage("Produto adicionado à sacola");
    setTimeout(() => {
      history.push("/");
    }, 1200);
  }

  return (
    <StyledShopCard>
      <p>
        Vendido e entregue por <span>RCHLO</span>
      </p>

      <div className="prices">
        <strong>R$ {price}</strong>
        <div>
          <span>Cartão de crédito 1x de R$ {price}</span>
          <span>Cartão de crédito 3x de R$ {installment()}</span>
        </div>
      </div>

      <button className="addBtn" onClick={() => handleAddToCart()}>
        ADICIONAR À SACOLA
      </button>

      <div className="deliver">
        <div className="deliver__info">
          <FiTruck size={20} color="#333" />
          <span>Consulte o frete</span>
          <input
            name="CEP"
            id="CEP"
            required
            pattern="\d{5}-\d{3}"
            placeholder="______-___"
          />
          <button>OK</button>
        </div>
        <p>Não sei meu cep</p>
      </div>
    </StyledShopCard>
  );
}
