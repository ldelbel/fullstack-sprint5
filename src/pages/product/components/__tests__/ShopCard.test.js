import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import { ShopCard } from "../ShopCard";
import { MemoryRouter, Route } from "react-router-dom";
import { act } from "react-dom/test-utils";
import MessageContext from "../../../../contexts/MessageContext";

const setMessage = () => ({});

describe("ShopCard component", () => {
  it("should render without crashing", () => {
    const { getByText } = render(
      <MemoryRouter initialEntries={["/product"]}>
        <MessageContext.Provider value={{ setMessage }}>
          <ShopCard price="300" />
        </MessageContext.Provider>
      </MemoryRouter>
    );

    expect(getByText("ADICIONAR À SACOLA")).toBeInTheDocument();
  });

  it("should get back to home when clicking on add to cart", async () => {
    let testLocation;

    await act(async () => {
      render(
        <MemoryRouter initialEntries={["/jaqueta"]}>
          <MessageContext.Provider value={{ setMessage }}>
            <ShopCard price="300" />
          </MessageContext.Provider>
          <Route
            path="*"
            render={({ location }) => {
              testLocation = location;
              return null;
            }}
          />
        </MemoryRouter>
      );
    });

    expect(testLocation.pathname).toBe("/jaqueta");

    fireEvent.click(screen.getByText("ADICIONAR À SACOLA"));

    await new Promise((r) => setTimeout(r, 2000));

    expect(testLocation.pathname).toBe("/");
  });

  it("should match snapshot", () => {
    const { container } = render(
      <MemoryRouter initialEntries={["/product"]}>
        <MessageContext.Provider value={{ setMessage }}>
          <ShopCard price="300" />
        </MessageContext.Provider>
      </MemoryRouter>
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
