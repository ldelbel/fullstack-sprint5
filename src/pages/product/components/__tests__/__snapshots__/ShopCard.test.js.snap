// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShopCard component should match snapshot 1`] = `
<div
  class="sc-bdnxRM frziFI"
>
  <p>
    Vendido e entregue por 
    <span>
      RCHLO
    </span>
  </p>
  <div
    class="prices"
  >
    <strong>
      R$ 
      300
    </strong>
    <div>
      <span>
        Cartão de crédito 1x de R$ 
        300
      </span>
      <span>
        Cartão de crédito 3x de R$ 
        100,00
      </span>
    </div>
  </div>
  <button
    class="addBtn"
  >
    ADICIONAR À SACOLA
  </button>
  <div
    class="deliver"
  >
    <div
      class="deliver__info"
    >
      <svg
        color="#333"
        fill="none"
        height="20"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        style="color: rgb(51, 51, 51);"
        viewBox="0 0 24 24"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          height="13"
          width="15"
          x="1"
          y="3"
        />
        <polygon
          points="16 8 20 8 23 11 23 16 16 16 16 8"
        />
        <circle
          cx="5.5"
          cy="18.5"
          r="2.5"
        />
        <circle
          cx="18.5"
          cy="18.5"
          r="2.5"
        />
      </svg>
      <span>
        Consulte o frete
      </span>
      <input
        id="CEP"
        name="CEP"
        pattern="\\\\d{5}-\\\\d{3}"
        placeholder="______-___"
        required=""
      />
      <button>
        OK
      </button>
    </div>
    <p>
      Não sei meu cep
    </p>
  </div>
</div>
`;
